File: haskell/src/Y2019/Day2.hs
module Y2019.Day2 (solve) where

import Data.List.Split (splitOn)

import Y2019.Intcode (runIntcodeEasy)

modIntcode :: [Int] -> Int -> Int -> [Int]
modIntcode (x:_:_:xs) n v = x:n:v:xs

p2 :: [Int] -> Int -> Int
p2 xs patt = head [ 100 * x + y | x <- [0..99], y <- [0..99], head (runIntcodeEasy $ modIntcode xs x y) == patt ]

solve :: String -> (String, String)
solve input = (part1, part2)
  where
    ns = map read $ splitOn "," input
    part1 = show . head . runIntcodeEasy $ modIntcode ns 12 2
    part2 = show $ p2 ns 19690720

File: clojure/src/y2023/day2.clj
(ns y2023.day2
  (:require
   [clojure.string :as s]))

(defn count-marbles [pattern s]
  (let [matches (re-seq (re-pattern (str "(\\d+) " pattern)) s)
        groups (map first matches)]
    (map read-string groups)))

(defn is-valid [s]
  (let [blues (count-marbles "blue" s)
        reds (count-marbles "red" s)
        greens (count-marbles "green" s)]
    (and (every? #(<= % 12) reds) (every? #(<= % 13) greens) (every? #(<= % 14) blues))))

(defn get-id [s]
  (read-string (re-find #"\d+" s)))

(defn power [s]
  (let [blues (count-marbles "blue" s)
        reds (count-marbles "red" s)
        greens (count-marbles "green" s)]
    (* (apply max blues) (apply max reds) (apply max greens))))

(defn solve [input]
  (let [lines (s/split-lines input)
        part1 (apply + (map get-id (filter is-valid lines)))
        part2 (apply + (map power lines))]
    [part1 part2]))
